CORTX-29592: Support deploy into non-default namespace

walterlopatka opened the pull request:
## Description
This change allows CORTX to be deployed into any namespace (including default).
* If the namespace does not exist it will be created.
* In the solution.example.yaml file the default namespace is now `cortx`

Note: This does not support multiple CORTX deployments into a single k8s cluster.  That will come later.

Note: I do not consider this to be a breaking change.  The default namespace in `solution.example.yaml` has changed from `default` to `cortx`.  But given an existing `solution.example.yaml` file, which already has a specific namespace specified.  This could possible change behavior of tools that automatically copy `solution.example.yaml` to generate the config file.


## Type of change
<!--
What type of change is this? Does it fix an issue, or is it new functionality? Check as many items
as necessary to accurately describe the change. If you are checking more than one of the items,
consider splitting it up into separate PRs if it makes sense.
-->
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds new functionality)
- [ ] Breaking change (bug fix or new feature that breaks existing functionality)
- [ ] Third-party dependency update
- [ ] Documentation additions or improvements
- [ ] Code quality improvements to existing code or test additions/updates

## Applicable issues
<!--
If this change directly fixes or is related to any existing GitHub or Jira issue, mention those
here. You can reference a GitHub issue using "#<issue number>". If this is related to a Seagate
internal issue (Jira), please reference the CORTX-NNNNN issue number.
-->
- This change is the first to address CORTX-29592


## How was this tested?
<!--
In-lieu of requiring automated tests for changes (we're working on that!), we are asking you to
provide a brief description of how this change was tested, especially any details specific to the
change.
-->
* Testing is in progress.

## Checklist
<!--
Place an 'x' in all the items that apply. You can also fill them out after the PR is submitted. This
serves as a reminder for what the maintainers will be looking for when reviewing the change.
-->

- [x] The change is tested and works locally.
- [x] New or changed settings in the solution YAML are documented clearly in the README.md file.
- [x] All commits are signed off and are in agreement with the [CORTX Community DCO and CLA policy](https://github.com/Seagate/cortx/blob/main/doc/dco_cla.md).

If this change addresses a CORTX Jira issue:

- [x] The title of the PR starts with the issue ID (e.g. `CORTX-XXXXX:`)

walterlopatka commented on k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-configmap/values.yaml:
```diff
@@ -45,16 +45,16 @@ clusterStorageSets: {}
   endpoints:
-    - tcp://consul-server.default.svc.cluster.local:8301
-    - http://consul-server.default.svc.cluster.local:8500
+    - tcp://consul-server.cortx.svc.cluster.local:8301
```
Perhaps these should be shortened to `tcp://consul-server:8301` (and so on).  As currently implemented, these are set by  `deploy-cortx-cloud.sh` as command line args to `helm install`.

walterlopatka commented on k8_cortx_cloud/cortx-cloud-3rd-party-pkg/templates/local-path-storage-template.yaml:
```diff
@@ -153,5 +153,3 @@ data:
       containers:
       - name: helper-pod
         image: <<.Values.rancher.helperPod.image>>
-        imagePullPolicy: IfNotPresent
```
Hmmmm... I have no idea why I would have removed this.  I'll restore it.

walterlopatka commented on k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-hax-svc.yaml:
```diff
@@ -1,3 +1,4 @@
+{{- if .Values.services.create -}}
```
`cortx-platform` is installed twice as part of the current `deploy-cortx-cloud.sh` script.  This was not an issue previously because "if namespace != default" guarded the first call to create the namespace.

walterlopatka replied:
Since posting this comment, there has been a change in that cortx-platform is now only helm-installed once, but this code remains until/unless we decide to remove the "create" field throughout this helm chart.

keithpine replied:
I think it's fine, we can fix it later.

walterlopatka commented on k8_cortx_cloud/deploy-cortx-cloud.sh:
```diff
@@ -1299,13 +1314,14 @@ if [[ "${num_worker_nodes}" -gt "${max_kafka_inst}" ]]; then
 fi
 
-if [[ (${#namespace_list[@]} -le 1 && "${found_match_nsp}" = true) || "${namespace}" == "default" ]]; then
+# WWW: Once this proves to be a good change I will delete the "if" statement.
```
Whoops... I should have removed this.

walterlopatka replied:
done

walterlopatka commented on k8_cortx_cloud/destroy-cortx-cloud.sh:
```diff
@@ -400,55 +378,49 @@ function delete3rdPartyPVs()
-        done <<< "${charts}"
-    fi
-}
+#function helmChartCleanup()
```
I commented out this entire function.  It's not obvious to me that it is needed, though I am not 100% sure yet.  I need to review more closely and make a final determination.

walterlopatka replied:
This has since been removed.

walterlopatka commented on k8_cortx_cloud/destroy-cortx-cloud.sh:
```diff
@@ -460,9 +432,9 @@ function deleteKubernetesPrereqs()
-    if [[ "${namespace}" != "default" ]]; then
-        uninstallHelmChart "cortx-ns-${namespace}"
-    fi
+    #if [[ "${namespace}" != "default" ]]; then
```
I'm not sure about this... I need to review this and consider it more carefully.

walterlopatka replied:
I restored this one.

walterlopatka commented on k8_cortx_cloud/destroy-cortx-cloud.sh:
```diff
@@ -515,22 +487,22 @@ for np in "${namespace_list[@]}"; do
 done
 
-if [[ (${#namespace_list[@]} -le 1 && "${found_match_np}" = true) || "${namespace}" == "default" ]]; then
+#if [[ (${#namespace_list[@]} -le 1 && "${found_match_np}" = true) || "${namespace}" == "default" ]]; then
```
Need to delete these commented out "if / fi" lines.

walterlopatka replied:
done

walterlopatka commented on k8_cortx_cloud/deploy-cortx-cloud.sh:
```diff
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+#set -x
```
Whoops... need to remove this.

walterlopatka replied:
removed

keithpine commented on k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-io-svc.yaml:
```diff
@@ -1,3 +1,4 @@
+{{- if .Values.services.create -}}
 {{- $svc := .Values.services.io -}}
 {{- if gt $svc.count 0 -}}
```
Could you make this fix too? (see #206)

```suggestion
{{- if gt (int $svc.count) 0 -}}
```

walterlopatka replied:
Yes, certainly!  I'll make sure that's in the net update.

walterlopatka replied:
This has been addressed.

osowski reviewed (commented):
Overall, these changes look great. Two comments that we can discuss more next week.

1. We can change the default create/delete practice to install Rancher with the `kubectl apply` command, instead of `kubectl create`, while also not deleting Rancher on a standard destroy script run. We can add a flag to the destroy script to force deletion of Rancher, but I don't think it's that much of an issue to leave Rancher deployed, as long as we use the `kubectl apply` method to create it.
2. We can remove the `namespace` element in each of the Helm Charts and instead of using `.Values.namespace`, we can use `.Release.Namespace` as it is always provided by Helm and we are explicitly passing in the `--namespace` to the `helm install` commands. I have already done this for CORTX-28968. We don't need to include `metadata.namespace` as a hard-coded namespace for objects, since the previous application of `.Release.Namespace` already does that for us. The less code we can write to do the same thing, the better.

keithpine commented on README.md:
```diff
@@ -204,7 +206,7 @@ All paths below are prefixed with `solution.` for fully-qualified naming and are
 | Name              | Description                                                                                                                  | Default Value |
 | ----------------- | ---------------------------------------------------------------------------------------------------------------------------- | ------------- |
-| `namespace`       | The Kubernetes namespace for CORTX Pods to be deployed in.                                                                   | `default`     |
+| `namespace`       | The Kubernetes namespace for CORTX Pods to be deployed in.                                                                   | `cortx`     |
```
Forgive me for commenting on something that's already existing, but two nitpicks:
* We deploy third party resources (Consul, etc.) into the same namespace, not just CORTX.
* It's also not just Pods.

keithpine reviewed (commented):
Looking good. I agree with Rick's comments.

keithpine commented:
Thought I'd mention, I had a manual install of consul in a non-default namespace. When I deployed CORTX into another namespace, the consul install failed. Not sure if it's related to my other namespace install or not.

walterlopatka commented:
I have addressed the comments I have received.  Specifically:
* I have removed the "namespace" values in the cortx helm charts, instead relying on .Release.Namespace to provide that information.
* I have changed "kubectl create" local-storage to "kubectl apply" local-storage, and removed the corresponding "delete" local storage.  (I also added a comment to the "undeploy" section of the doc noting how one can delete the local storage.  I thought this was better than adding a bash option.)
* I addressed Keith's legit comment about the namespace documentation

I have tested these changes, but I will do more testing early next week before merging.

(I have not yet looked into Keith's comment about deploying with consul already existing in another namespace.  I'll look at that on Monday.)

keithpine commented on README.md:
```diff
@@ -204,7 +215,7 @@ All paths below are prefixed with `solution.` for fully-qualified naming and are
 | Name              | Description                                                                                                                  | Default Value |
 | ----------------- | ---------------------------------------------------------------------------------------------------------------------------- | ------------- |
-| `namespace`       | The Kubernetes namespace for CORTX Pods to be deployed in.                                                                   | `default`     |
+| `namespace`       | The Kubernetes namespace that all CORTX-related resource will be deployed into.
```
```suggestion
| `namespace`       | The Kubernetes namespace that all CORTX-related resources will be deployed into.
```

keithpine commented on k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-io-svc.yaml:
```diff
@@ -9,7 +10,7 @@ apiVersion: v1
 metadata:
   name: {{ printf "%s-%d" $svc.name $i | trunc 63 | trimSuffix "-" }}
-  namespace: {{ $root.Values.namespace.name }}
+  namespace: {{ $.Release.Namespace }}
```
```suggestion
  namespace: {{ $root.Release.Namespace }}
```

keithpine commented on k8_cortx_cloud/deploy-cortx-cloud.sh:
```diff
@@ -420,15 +395,18 @@ function deployConsul()
-    kubectl patch serviceaccount/consul-client -p '{"automountServiceAccountToken":false}'
-    kubectl patch serviceaccount/consul-server -p '{"automountServiceAccountToken":false}'
+    kubectl patch serviceaccount/consul-client -p '{"automountServiceAccountToken":false}' \
+                                               -n "${namespace}"
```
```suggestion
                                               --namespace "${namespace}"
```

There's not a consistent use of `-n` and `--namespace` in this script already. My preference is to always use long options in scripts since they are self documenting.

keithpine commented on k8_cortx_cloud/deploy-cortx-cloud.sh:
```diff
@@ -420,15 +395,18 @@ function deployConsul()
+    kubectl patch serviceaccount/consul-client -p '{"automountServiceAccountToken":false}' \
+                                               -n "${namespace}"
+    kubectl patch serviceaccount/consul-server -p '{"automountServiceAccountToken":false}' \
+                                               -n "${namespace}"
```
```suggestion
                                               --namespace "${namespace}"
```

keithpine commented on k8_cortx_cloud/deploy-cortx-cloud.sh:
```diff
@@ -420,15 +395,18 @@ function deployConsul()
     # Rollout a new deployment version of Consul pods to use updated Service Account settings
-    kubectl rollout restart statefulset/consul-server
-    kubectl rollout restart daemonset/consul-client
+    kubectl rollout restart statefulset/consul-server -n "${namespace}"
+    kubectl rollout restart daemonset/consul-client -n "${namespace}"
```
```suggestion
    kubectl rollout restart statefulset/consul-server --namespace "${namespace}"
    kubectl rollout restart daemonset/consul-client --namespace "${namespace}"
```

keithpine commented on k8_cortx_cloud/deploy-cortx-cloud.sh:
```diff
@@ -928,9 +910,9 @@ function waitForAllDeploymentsAvailable()
     # Initial wait
     FAIL=0
-    if ! kubectl wait --for=condition=available --timeout="${TIMEOUT}" "$@"; then
+    if ! kubectl wait --for=condition=available --timeout="${TIMEOUT}" -n "${NAMESPACE}" "$@"; then
```
```suggestion
    if ! kubectl wait --for=condition=available --timeout="${TIMEOUT}" --namespace "${NAMESPACE}" "$@"; then
```

keithpine commented on k8_cortx_cloud/deploy-cortx-cloud.sh:
```diff
@@ -928,9 +910,9 @@ function waitForAllDeploymentsAvailable()
+    if ! kubectl wait --for=condition=available --timeout="${TIMEOUT}" -n "${NAMESPACE}" "$@"; then
         # Secondary wait
-        if ! kubectl wait --for=condition=available --timeout="${TIMEOUT}" "$@"; then
+        if ! kubectl wait --for=condition=available --timeout="${TIMEOUT}" -n "${NAMESPACE}" "$@"; then
```
```suggestion
        if ! kubectl wait --for=condition=available --timeout="${TIMEOUT}" --namespace "${NAMESPACE}" "$@"; then
```

keithpine commented on k8_cortx_cloud/deploy-cortx-cloud.sh:
```diff
@@ -1366,12 +1334,12 @@ if [[ ${deployment_type} != "data-only" ]]; then
    Default IAM access key: ${data_service_default_user}
    Default IAM secret key is accessible via:
-       kubectl get secrets/${cortx_secret_name} --template={{.data.s3_auth_admin_secret}} | base64 -d
+       kubectl get secrets/${cortx_secret_name} -n ${namespace} --template={{.data.s3_auth_admin_secret}} | base64 -d
```
```suggestion
       kubectl get secrets/${cortx_secret_name} --namespace ${namespace} --template={{.data.s3_auth_admin_secret}} | base64 -d
```

keithpine commented on k8_cortx_cloud/deploy-cortx-cloud.sh:
```diff
@@ -1366,12 +1334,12 @@ if [[ ${deployment_type} != "data-only" ]]; then
    Default control username: ${control_service_default_user}
    Default control password is accessible via:
-       kubectl get secrets/${cortx_secret_name} --template={{.data.csm_mgmt_admin_secret}} | base64 -d"
+       kubectl get secrets/${cortx_secret_name} -n ${namespace} --template={{.data.csm_mgmt_admin_secret}} | base64 -d"
```
```suggestion
       kubectl get secrets/${cortx_secret_name} --namespace ${namespace} --template={{.data.csm_mgmt_admin_secret}} | base64 -d"
```

keithpine commented on k8_cortx_cloud/destroy-cortx-cloud.sh:
```diff
@@ -336,7 +320,7 @@ function waitFor3rdPartyToTerminate()
         while IFS= read -r line; do
             count=$(( count + 1 ))
-        done < <(kubectl get pods | grep -e kafka -e zookeeper -e openldap -e consul 2>&1)
+        done < <(kubectl get pods -n "${namespace}" | grep -e kafka -e zookeeper -e openldap -e consul 2>&1)
```
```suggestion
        done < <(kubectl get pods --namespace "${namespace}" | grep -e kafka -e zookeeper -e openldap -e consul 2>&1)
```

keithpine commented on k8_cortx_cloud/destroy-cortx-cloud.sh:
```diff
@@ -349,48 +333,24 @@ function delete3rdPartyPVCs()
         printf "Removing %s\n" "${volume_claim}"
         if [[ "${force_delete}" == "--force" || "${force_delete}" == "-f" ]]; then
-            kubectl patch pvc "${volume_claim}" -p '{"metadata":{"finalizers":null}}'
+            kubectl patch pvc -n "${namespace}" "${volume_claim}" -p '{"metadata":{"finalizers":null}}'
```
```suggestion
            kubectl patch pvc --namespace "${namespace}" "${volume_claim}" -p '{"metadata":{"finalizers":null}}'
```

keithpine commented on k8_cortx_cloud/destroy-cortx-cloud.sh:
```diff
@@ -349,48 +333,24 @@ function delete3rdPartyPVCs()
+            kubectl patch pvc -n "${namespace}" "${volume_claim}" -p '{"metadata":{"finalizers":null}}'
         fi
-        kubectl delete pvc "${volume_claim}"
+        kubectl delete pvc -n "${namespace}" "${volume_claim}"
```
```suggestion
        kubectl delete pvc --namespace "${namespace}" "${volume_claim}"
```

keithpine reviewed (approved):
Just some nitpick comments, fix them if you want, otherwise, LGTM.

walterlopatka commented:
Thanks @keithpine for the thorough review.  I have accepted all those suggestions.  I think it's ready to go.  I'll give @osowski a shot at reviewing this later today if he wants.

osowski reviewed (approved):
lgtm!
